core/services/relay/evm/evmtesting/chain_reader_interface_tester.go

```diff
@@ -0,0 +1,330 @@
+	it.IncNonce()
+	it.AwaitTx(t, tx)
+	it.dirtyContracts = true
+}
+
+func (it *EvmChainReaderInterfaceTester[T]) GetAuthWithGasSet(t T) *bind.TransactOpts {
+	gasPrice, err := it.client.SuggestGasPrice(it.Helper.Context(t))
+	require.NoError(t, err)
+	extra := new(big.Int).Mul(gasPrice, big.NewInt(it.Helper.GasPriceBufferPercent()))
+	extra = extra.Div(extra, big.NewInt(100))
+	it.auth.GasPrice = gasPrice.Add(gasPrice, extra)
+	return it.auth
+}
+
+func (it *EvmChainReaderInterfaceTester[T]) IncNonce() {
+	if it.auth.Nonce == nil {
+		it.auth.Nonce = big.NewInt(1)
+	} else {
+		it.auth.Nonce = it.auth.Nonce.Add(it.auth.Nonce, big.NewInt(1))
+	}
+}
+
+func (it *EvmChainReaderInterfaceTester[T]) AwaitTx(t T, tx *evmtypes.Transaction) {
+	ctx := it.Helper.Context(t)
+	receipt, err := bind.WaitMined(ctx, it.client, tx)
+	require.NoError(t, err)
+	require.Equal(t, evmtypes.ReceiptStatusSuccessful, receipt.Status)
+}
+
+func (it *EvmChainReaderInterfaceTester[T]) deployNewContracts(t T) {
+	// First test deploy both contracts, otherwise only deploy contracts if cleanup decides that we need to.
+	if it.address == "" {
+		it.address = it.deployNewContract(t)
+		it.address2 = it.deployNewContract(t)
+	} else if it.evmTest == nil {
+		it.address = it.deployNewContract(t)
+		it.dirtyContracts = false
+	}
+}
+
+func (it *EvmChainReaderInterfaceTester[T]) deployNewContract(t T) string {
+	// 105528 was in the error: gas too low: have 0, want 105528
+	// Not sure if there's a better way to get it.
+	it.auth.GasLimit = 10552800
+
+	address, tx, ts, err := chain_reader_tester.DeployChainReaderTester(it.GetAuthWithGasSet(t), it.Helper.Backend())
+	require.NoError(t, err)
+	it.Helper.Commit()
+	if it.evmTest == nil {
+		it.evmTest = ts
+	}
+
+	it.IncNonce()
+	it.AwaitTx(t, tx)
+	return address.String()
+}
+
+func (it *EvmChainReaderInterfaceTester[T]) MaxWaitTimeForEvents() time.Duration {
+	return it.Helper.MaxWaitTimeForEvents()
+}
+
+func ConvertOracleIDs(oracleIDs [32]commontypes.OracleID) [32]byte {
+	convertedIds := [32]byte{}
+	for i, id := range oracleIDs {
+		convertedIds[i] = byte(id)
+	}
+	return convertedIds
+}
+
+func ConvertAccounts(accounts [][]byte) []common.Address {
+	convertedAccounts := make([]common.Address, len(accounts))
+	for i, a := range accounts {
+		convertedAccounts[i] = common.Address(a)
+	}
+	return convertedAccounts
+}
+
+func ToInternalType(testStruct TestStruct) chain_reader_tester.TestStruct {
+	return chain_reader_tester.TestStruct{
+		Field:          *testStruct.Field,
+		DifferentField: testStruct.DifferentField,
+		OracleId:       byte(testStruct.OracleID),
+		OracleIds:      ConvertOracleIDs(testStruct.OracleIDs),
+		Account:        common.Address(testStruct.Account),
+		Accounts:       ConvertAccounts(testStruct.Accounts),
+		BigField:       testStruct.BigField,
+		NestedStruct:   MidToInternalType(testStruct.NestedStruct),
+	}
+}
+
+func MidToInternalType(m MidLevelTestStruct) chain_reader_tester.MidLevelTestStruct {
```

Comment:
Also a candidate for a method

Reply:
The type lives in chainlin-common and doesn't know about EVM types.  The chain_reader_tester type is generated from gethwrappers.

I could put the func next to the generated code, but wasn't sure if that would be best practice for a test like this.